pallets/phala/src/stakepool.rs

```diff
@@ -471,6 +559,95 @@ pub mod pallet {
 				NewRewards::<T>::insert(&worker, BalanceOf::<T>::zero());
 			}
 		}
+
+		/// Try to withdraw specific amount from pool, would be delayed if the free stake is not enough.
+		fn try_withdraw(
+			pool_info: &mut PoolInfo<T::AccountId, BalanceOf<T>>,
+			user_info: &mut UserStakeInfo<T::AccountId, BalanceOf<T>>,
+			amount: BalanceOf<T>,
+		) {
+			Self::update_pool(pool_info.pid);
+
+			// enough free stake, withdraw directly
+			if pool_info.free_stake >= amount {
+				pool_info.free_stake = pool_info.free_stake.saturating_sub(amount);
+				pool_info.total_stake = pool_info.total_stake.saturating_sub(amount);
+				user_info.amount = user_info.amount.saturating_sub(amount);
+			} else {
+				let now = <T as registry::Config>::UnixTime::now()
+					.as_secs()
+					.saturated_into::<u64>();
+				// all of the free_stake would be withdrew back to user
+				let unwithdraw_amount = amount.saturating_sub(pool_info.free_stake);
+				pool_info.free_stake = Zero::zero();
+				pool_info.total_stake = pool_info.total_stake.saturating_sub(pool_info.free_stake);
+				user_info.amount = user_info.amount.saturating_sub(unwithdraw_amount);
+
+				// case some locked asset has not been withdraw(unlock) to user, add it to withdraw queue.
+				// when pool has free stake again, the withdraw would be handled
+				pool_info.withdraw_queue.push_back(WithdrawInfo {
+					user: user_info.user.clone(),
+					amount: unwithdraw_amount,
+					start_time: now,
+				});
+			}
+			user_info.user_debt =
+				user_info.amount * pool_info.pool_acc / 10u32.pow(6).saturated_into();
+
+			// update the lock balance of user
+			Self::update_lock(user_info.user.clone(), user_info.amount);
+		}
+
+		fn try_handle_waitting_withdraw(pool_info: &mut PoolInfo<T::AccountId, BalanceOf<T>>) {
+			while pool_info.free_stake > Zero::zero() {
+				if pool_info.withdraw_queue.is_empty() {
+					break;
+				}
+
+				if let Some(mut withdraw_info) = pool_info.withdraw_queue.pop_front() {
+					let info_key = (pool_info.pid.clone(), withdraw_info.user.clone());
+					let mut user_info = Self::staking_info(&info_key).unwrap();
+
+					if pool_info.free_stake <= withdraw_info.amount {
+						pool_info.free_stake = Zero::zero();
+						pool_info.total_stake =
+							pool_info.total_stake.saturating_sub(pool_info.free_stake);
+						withdraw_info.amount =
+							withdraw_info.amount.saturating_sub(pool_info.free_stake);
+
+						// push front the updated withdraw info
+						pool_info.withdraw_queue.push_front(withdraw_info);
```

Comment:
When `free_stake == withdraw_info.amount`, the withdraw request has been fulfilled. We don't want to push it back to the queue head

Reply:
Yeah, logic mistake